src/Adapter/Product/CommandHandler/UpdateProductStatusHandler.php

```diff
@@ -66,19 +66,22 @@ public function __construct(
      */
     public function handle(UpdateProductStatusCommand $command)
     {
-        $product = $this->productRepository->get($command->getProductId());
+        $shopConstraint = $command->getShopConstraint();
+        $product = $this->productRepository->getByShopConstraint($command->getProductId(), $shopConstraint);
         $initialState = (bool) $product->active;
         $product->active = $command->getEnable();
         $this->productRepository->partialUpdate(
             $product,
             ['active'],
+            $shopConstraint,
             CannotUpdateProductException::FAILED_UPDATE_STATUS
         );
 
         // If status changed we need to update its indexes (we check if it is necessary because index build can be
-        // an expensive operation).
-        if ($initialState !== $command->getEnable()) {
-            $this->productIndexationUpdater->updateIndexation($product);
+        // an expensive operation). We cannot have this optimization check in multi-shop context, because product is loaded from a single shop,
+        // so it would end up checking one shop product and leaving all other shops unhandled
+        if (!$shopConstraint->getShopId() || $initialState !== $command->getEnable()) {
```

Comment:
Same suggestion as in the options handler The multishop use cases are quite complex sometimes so it's better if the code is really explicit about its intention even if it results in more code sometimes

Reply:
I agreed with options, just to avoid additional visibility check before.
But in this case seems ok for me.
2 lines of code, simple and clean + has a comment.
I think you are overdoing it here :smile:

Reply:
If this PR takes a little bit longer to merge, then after like a week, someone will ask me to "optimize this code because it can be written in 2 lines instead of 6". :sweat_smile: 
:ok: